### ISSUE TITLE
Placebo masks are not de-duplicated: with few geos the null collapses to a point mass and `p` becomes `1/(1 + n_top_placebos)` regardless of the data

### ISSUE THREAD

**Issue by PelayoPuerta** [google/meridian-geox#15]
## Summary

`_get_placebo_masks` builds the placebo null from `n_top_placebos` (default 500) masks taken from a pool that is generated **with replacement and never de-duplicated**. When the number of geos is small, the pool contains only a handful of distinct masks, and taking the top `n_top_placebos` by out-of-sample R² returns **500 copies of the single best-fitting one**.

The null is then a point mass, and every statistic derived from it degenerates:

- `p_value` is `(1 + n_extreme) / (1 + n_placebo)` with `n_placebo = 500` counting copies, so it collapses to `1/501 = 0.001996` (or to `1.0`) **independently of the data**;
- `standard_deviation` is `std(t_placebo) * rmse` = `0.0`;
- the confidence interval is a quantile of a constant, so it has zero width and generally does **not contain the point estimate**.

Reproduced below on synthetic **A/A** data (no treatment effect anywhere) with 13 geos and factory-default `AnalysisConfig`: `analyze()` reports `p = 0.001996`. That is a false positive with probability 1, not with probability α.

Version: `meridian-geox` **1.0.1**, and `main` @ `0a1dd0f730ab8170e4e46646f3bf69400d57d021` — `analysis.py`, `generate_candidates.py` and `methodology/util.py` are byte-identical between 0.1.1 and 1.0.1, so the 1.0.x releases did not touch this path.

## Reproduction

```python
import datetime as dt
import numpy as np, pandas as pd, jax
import meridian_geox as mg
from meridian_geox import api, analysis as mg_analysis

N_GEOS, N_DAYS, TEST = 13, 120, 21
geos = [f"geo_{i:02d}" for i in range(N_GEOS)]
rng = np.random.default_rng(0)
dates = pd.date_range("2026-01-01", periods=N_DAYS, freq="D")
dow = 1 + 0.15 * np.sin(np.arange(N_DAYS) * 2 * np.pi / 7)          # weekly seasonality
df = pd.concat([pd.DataFrame({"date": dates, "location": g,
        "conversions": lv * dow * (1 + 0.02 * rng.standard_normal(N_DAYS))})
    for g, lv in zip(geos, np.linspace(100, 220, N_GEOS))], ignore_index=True)
# NO treatment effect is applied anywhere: this is pure A/A.

dcfg = api.DesignConfig(experiment_duration=dt.timedelta(days=TEST),
    experiment_types=api.ExperimentType.GO_DARK, methodology=api.Methodology.TBR,
    geo_assignment_rule=api.GeoAssignmentRule.RANDOM, cell_count=1,
    alpha=0.10, test_type=api.TestType.TWO_SIDED, seed=42)
design = api.Design(designs={"cell_1": api.PerCellDesign(treatment_geos={geos[0]},
        minimum_detectable_effect=float("nan"), design_implied_cpic=float("nan"),
        p_value=float("nan"), budget=0.0)},
    control_geos=set(geos[1:]), excluded_geos=set(), excluded_dates=set(),
    design_config=dcfg, constraints=api.Constraints(max_conversions_percent=0.20), data=df)
acfg = api.AnalysisConfig(design=design,                       # factory defaults
    analysis_start_date=df.date.max() - pd.Timedelta(days=TEST - 1),
    analysis_end_date=df.date.max())

lift = mg.analyze(df, acfg).results["cell_1"].lift
print(f"estimate {lift.point_estimate:.2f}  CI [{lift.lower_bound:.2f}, {lift.upper_bound:.2f}]"
      f"  sd {lift.standard_deviation:.4f}  p {lift.p_value:.6f}")

masks = np.array(mg_analysis._get_placebo_masks(design,                       # diagnostic
    mg_analysis._get_treatment_mask(df, design), dcfg, acfg,
    jax.random.fold_in(jax.random.key(dcfg.seed), 12345)))
print(f"placebo masks used: {len(masks)}, distinct: {len(np.unique(masks, axis=0))}")
```

```
estimate -1.34  CI [0.00, 0.00]  sd 0.0000  p 0.001996
placebo masks used: 500, distinct: 1
```

The point estimate is `-1.34` and the reported 90% CI is `[0.00, 0.00]`, which does not contain it.

## What is happening

1. [`generate_candidates._generate_random_masks`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/generate_candidates.py#L583) draws `n_designs` masks independently (`jax.random.uniform` scores → `argsort`), i.e. **with replacement over the space of masks**. [`get_random_candidates`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/generate_candidates.py#L627) filters them by volume and trims to `n_designs`, but never de-duplicates.
2. With 12 control geos and `max_conversions_percent = 0.20`, `n_treated = int(12 * 0.20) = 2`, so only `C(12, 2) = 66` masks exist. Measured on the 100,000 raw draws: **66 distinct masks, each appearing 1,410-1,620 times** (mean 1,515). After the volume filter, 60 distinct masks remain in the pool.
3. [`analysis._get_placebo_masks`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/analysis.py#L228-L233) sorts by R² descending and keeps the first `n_top_placebos`. Since the best-fitting mask alone occurs ~1,500 times, the top 500 are **500 copies of it**.
4. [`methodology/util.compute_studentized_p_value`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/methodology/util.py#L43-L52) uses `n_placebo = len(t_placebo)`, which counts copies. [`compute_se`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/methodology/util.py#L62) and [`compute_cis`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/methodology/util.py#L66-L81) take the std and the quantiles of that constant vector. The same pattern appears again in [`get_percent_lift`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/methodology/util.py#L127-L162).

## The reported p-value is a function of a config parameter, not of the data

Same synthetic panel, varying the true effect applied to the treated geo and `n_top_placebos`:

| true effect | `n_top_placebos` | masks (distinct) | estimate | sd | p |
|---|---|---|---|---|---|
| 0% | 500 | 500 (1) | -1.34 | 0.0 | 0.001996 |
| -10% | 500 | 500 (1) | 208.65 | 0.0 | 0.001996 |
| -30% | 500 | 500 (1) | 628.63 | 0.0 | 0.001996 |
| 0% | 50 | 50 (1) | -1.34 | 0.0 | 0.019608 |
| -10% | 50 | 50 (1) | 208.65 | 0.0 | 0.019608 |

`p` equals `1/(1 + n_top_placebos)` in every row. A 30% effect and no effect at all receive the same p-value; changing `n_top_placebos` changes the p-value by an order of magnitude.

## When it bites

Number of distinct masks among the `n_top_placebos = 500` kept, A/A data, `max_conversions_percent = 0.20`:

| geos | control geos | placebo treated (`int(n·0.2)`) | masks that exist | distinct in pool | distinct in top 500 | sd | p |
|---|---|---|---|---|---|---|---|
| 13 | 12 | 2 | 66 | 60 | **1** | 0.00 | 0.001996 |
| 16 | 15 | 3 | 455 | 229 | 2 | 8.68 | 0.325 |
| 20 | 19 | 3 | 969 | 962 | 5 | 16.96 | 0.812 |
| 25 | 24 | 4 | 10,626 | 10,411 | 51 | 18.51 | 0.768 |
| 30 | 29 | 5 | 118,755 | 66,580 | 354 | 15.54 | 0.828 |

So the failure is silent and specific to small panels — which is the common case for a country split into administrative regions (Spain has 17 autonomous communities; a single-region go-dark leaves ~12-16 usable controls).

## The existing guard rail counts the wrong thing

`min_placebo_count_warning` (default 100) and `min_placebo_count_error` (default 10) are checked in [`util.filter_by_r2`](https://github.com/google/meridian-geox/blob/0a1dd0f730ab8170e4e46646f3bf69400d57d021/meridian_geox/util.py#L79) against the number of candidates passing the R² threshold, **with duplicates**. In the reproduction above that count is **100,000**, so no warning is emitted, while the number of distinct masks is 60 and the number used is 1.

## `run_design` is affected too

The same generator feeds design candidate generation. On the same synthetic panel (13 geos, `HOLDBACK`, `RANDOM` assignment, factory defaults), `run_design` returns 10 designs and all 10 are the same pair of geos:

```
returned designs: 10 | distinct treatment sets: 1
   x10  ('geo_05', 'geo_06')
```

## Suggested fix

De-duplicate the candidate masks before they are scored and ranked. Patching `get_random_candidates` to return `jnp.array(np.unique(np.array(candidates), axis=0))` is enough to restore sensible behaviour on the reproduction above:

| true effect | as shipped | pool de-duplicated |
|---|---|---|
| 0% | 500 placebos (1 distinct), sd 0.00, **p = 0.001996** | 60 placebos, sd 10.65, **p = 0.836** |
| -10% | 500 (1), sd 0.00, p = 0.001996 | 60, sd 10.65, p = 0.016 |
| -30% | 500 (1), sd 0.00, p = 0.001996 | 60, sd 10.65, p = 0.016 |

(`0.016 = 1/61` is the resolution floor of a 60-placebo permutation test, which is the honest answer for a panel this size.)

Two things that would help beyond the de-duplication itself:

- when the number of **distinct** masks is below `n_top_placebos`, say so — the resolution floor of the test is `1/(1 + n_distinct)` and the user needs to know it. The current warning thresholds should count distinct masks, not rows.
- when the mask space is small (`C(n_control, n_treated)` below, say, a few thousand), enumerating it exactly is cheaper than sampling 100,000 times and gives the exact permutation null.

## Environment

`meridian-geox` 1.0.1 · Python 3.13.15 · macOS arm64 · jax 0.11.1 · numpy 2.5.3 · pandas 3.0.5 · scipy 1.18.1

**Comment by PelayoPuerta** [comment]
Follow-up: the same defect reaches **`run_design`**, and there it shows that de-duplicating the candidate pool is not sufficient on its own.

## `run_design` returns `design_output_count` designs that are mostly the same design

Synthetic panels, `STRATIFIED_SAMPLING` (the rule the design side uses), `HOLDBACK`, `n_candidates=20000`, `n_ranked_candidates=100`, `design_output_count=10`, `min_r2=0.30`, `seed=42`:

| geos | pool drawn | pool **distinct** | most frequent mask | designs returned | **distinct** treatment sets |
|---|---|---|---|---|---|
| 10 | 20,000 | 112 | 1,494× | 10 | **1** |
| 13 | 20,000 | 441 | 425× | 10 | **2** |
| 16 | 20,000 | 1,425 | 390× | 10 | **3** |
| 20 | 20,000 | 5,292 | 512× | 10 | **5** |
| 30 | 20,000 | 15,503 | 155× | 10 | **7** |

With 10 geos the user asks for ten designs and gets one, listed ten times with ten different `design_id`s. And it never reaches ten distinct, not even with 30 geos and 15,503 distinct masks in the pool.

## Why this is a different lesson from the placebo case

In the placebo case the pool itself was degenerate (60 distinct masks out of 100,000 draws), so de-duplicating the pool fixes it. **Here the pool is diverse** — 15,503 distinct masks in the last row — and the output still collapses, because the ranking step keeps the first `n_ranked_candidates` rows of a list that contains **repeats**: a mask that was drawn 512 times occupies 512 consecutive positions once sorted by score. `design_output_count=10` then takes the top 10 of that.

So the fix has to be **de-duplication before the ranking**, not only at generation: whatever selects the top-k should be selecting among *distinct* masks. That covers both symptoms with one change.

## Reproduction

```python
import datetime as dt, warnings
import numpy as np, pandas as pd
import meridian_geox as mg
from meridian_geox import api, design as mg_design, generate_candidates as gc
warnings.filterwarnings("ignore")

N_DAYS, TEST = 165, 28

def panel(n_geos, seed=0):
    rng = np.random.default_rng(seed)
    geos = [f"geo_{i:02d}" for i in range(n_geos)]
    dates = pd.date_range("2026-01-01", periods=N_DAYS, freq="D")
    dow = 1 + 0.15 * np.sin(np.arange(N_DAYS) * 2 * np.pi / 7)
    return pd.concat([pd.DataFrame({"date": dates, "location": g,
            "conversions": lv * dow * (1 + 0.20 * rng.standard_normal(N_DAYS))})
        for g, lv in zip(geos, np.geomspace(20, 900, n_geos))], ignore_index=True)

seen = {}                                     # diagnostic: what the generator handed over
_orig = gc.get_stratified_sampling_candidates
def spy(**kw):
    out = _orig(**kw); a = np.array(out)
    u, c = np.unique(a, axis=0, return_counts=True)
    seen.update(n=len(a), uniq=len(u), top=int(c.max()))
    return out
gc.get_stratified_sampling_candidates = spy
mg_design.generate_candidates.get_stratified_sampling_candidates = spy

for n in (10, 13, 16, 20, 30):
    cfg = api.DesignConfig(
        experiment_duration=dt.timedelta(days=TEST), experiment_types=api.ExperimentType.HOLDBACK,
        geo_assignment_rule=api.GeoAssignmentRule.STRATIFIED_SAMPLING, cell_count=1,
        alpha=0.10, power=0.8, test_type=api.TestType.ONE_SIDED,
        cost_per_incremental_conversion=60.0, design_output_count=10,
        n_candidates=20000, n_ranked_candidates=100, min_r2=0.30, seed=42)
    ds = mg.run_design(panel(n), cfg, api.Constraints(max_conversions_percent=0.20))
    sets = [tuple(sorted(d.designs["cell_1"].treatment_geos)) for d in ds.designs.values()]
    print(f"{n:3d} geos · pool {seen['uniq']:6d} distinct (top mask {seen['top']:5d}×)"
          f" · returned {len(sets)}, distinct {len(set(sets))}")
```

## One more symptom worth noting

`min_r2` stops discriminating once this happens. On a 19-geo panel of ours, `min_r2=0.30` and `min_r2=0.05` return **the same single design** with identical metrics — which is what you would expect if only one distinct candidate ever reaches the output, but reads to the user as "lowering the threshold changed nothing, so 0.30 must already be permissive".

Same version as the issue above: `meridian-geox` 1.0.1, `main` @ `0a1dd0f`.

**Comment by AdimDrewnik** [comment]
Just out of curiosity, which tool did you use to find this issue and make this description?

**Comment by PelayoPuerta** [comment]
> Just out of curiosity, which tool did you use to find this issue and make this description?

Wow, I didn't know this was published. It was a private working session with Claude Code, and I had no idea it had published the issue on its own

**Comment by AdimDrewnik** [comment]
Interesting :) Which particular model did you use then? Did it find any other issues with GeoX?

**Comment by PelayoPuerta** [comment]
I was working with Opus 5. I think something else came up, though I don't remember it right now, I'd have to check
